File: src/adc.cpp
#include "adc.hpp"

#include <algorithm>
#include <cmath>

namespace volt {

SimulatedADC::SimulatedADC(const Config& cfg) : cfg_(cfg) {}

int SimulatedADC::max_level() const {
    int n = cfg_.n_bits_adc;
    if (n < 1) {
        return 0;
    }
    return (1 << n) - 1;
}

float SimulatedADC::i_step() const {
    int denom = max_level();
    if (denom <= 0) {
        return cfg_.I_range;
    }
    return cfg_.I_range / static_cast<float>(denom);
}

int SimulatedADC::quantize(float current) {
    float step = i_step();
    if (step <= 0.0f) {
        return 0;
    }
    float shifted = current - cfg_.I_min;
    float raw = std::floor(shifted / step);
    int level = static_cast<int>(raw);
    level = std::clamp(level, 0, max_level());
    return level;
}

float SimulatedADC::reconstruct(int level) const {
    level = std::clamp(level, 0, max_level());
    return cfg_.I_min + static_cast<float>(level) * i_step();
}

float SimulatedADC::level_to_dac_normalized(int level) const {
    const int mx = max_level();
    if (mx <= 0) {
        return 0.0f;
    }
    level = std::clamp(level, 0, mx);
    return static_cast<float>(level) / static_cast<float>(mx);
}

}  // namespace volt

File: crates/volt-core/src/adc.rs
use crate::config::Config;

pub struct SimulatedAdc {
    cfg: Config,
}

impl SimulatedAdc {
    pub fn new(cfg: Config) -> Self {
        Self { cfg }
    }

    pub fn max_level(&self) -> i32 {
        let n = self.cfg.n_bits_adc;
        if n < 1 {
            return 0;
        }
        (1 << n) - 1
    }

    pub fn i_step(&self) -> f32 {
        let denom = self.max_level();
        if denom <= 0 {
            return self.cfg.i_range;
        }
        self.cfg.i_range / denom as f32
    }

    pub fn quantize(&self, current: f32) -> i32 {
        let step = self.i_step();
        if step <= 0.0 {
            return 0;
        }
        let shifted = current - self.cfg.i_min;
        let raw = (shifted / step).floor() as i32;
        raw.clamp(0, self.max_level())
    }

    pub fn reconstruct(&self, level: i32) -> f32 {
        let level = level.clamp(0, self.max_level());
        self.cfg.i_min + level as f32 * self.i_step()
    }

    pub fn level_to_dac_normalized(&self, level: i32) -> f32 {
        let mx = self.max_level();
        if mx <= 0 {
            return 0.0;
        }
        let level = level.clamp(0, mx);
        level as f32 / mx as f32
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn adc_bipolar_window() {
        let mut cfg = Config::default();
        cfg.n_bits_adc = 8;
        let adc8 = SimulatedAdc::new(cfg.clone());
        let mid = cfg.i_min + cfg.i_range * 0.5;
        let l8 = adc8.quantize(mid);
        assert!((l8 - 127).abs() <= 1);

        cfg.n_bits_adc = 4;
        let adc4 = SimulatedAdc::new(cfg.clone());
        let l4 = adc4.quantize(mid);
        assert!((l4 - 7).abs() <= 1);
    }
}
